test/functional/cursor.test.js

```diff
@@ -3901,6 +3901,43 @@ describe('Cursor', function () {
     });
   });
 
+  // NODE-2035
+  it('should propagate error when exceptions are thrown from an awaited forEach call', function () {
+    const configuration = this.configuration;
+    const client = configuration.newClient({ w: 1 }, { maxPoolSize: 1 });
+    const docs = [{ a: 1 }, { a: 2 }, { a: 3 }];
+    let collection;
+    let db;
+    return client
+      .connect()
+      .then(() => {
+        db = client.db(configuration.db);
+        collection = db.collection('cursor_session_tests2');
+        return collection.insertMany(docs);
+      }, console.error)
+      .then(() => {
+        const cursor = collection.find({});
+        async function testAsync() {
+          let promiseResult;
+          try {
+            promiseResult = await cursor
+              .forEach(() => {
+                throw new Error('FAILURE IN FOREACH CALL');
+              })
+              .catch(err => {
+                expect(err.message).to.deep.equal('FAILURE IN FOREACH CALL');
+              });
+          } catch (err) {
+            expect(err).to.be.undefined;
+          }
+          expect(promiseResult).to.be.undefined;
+          cursor.close();
+          client.close();
```

Comment:
If the suggested change above is made make sure you still return the client.close() promise.